src/project.coffee

```diff
@@ -96,11 +110,45 @@ class Project extends Model
 
   # Public: Get an {Array} of {GitRepository}s associated with the project's
   # directories.
+  #
+  # This method will be removed in 2.0 because it does synchronous I/O.
+  # Prefer the following, which evaluates to a Promise that resolves to an
+  # {Array} of {Repository} objects:
+  # ```
+  # project.getDirectories().map(
+  #     project.repositoryForDirectory.bind(project))
+  # ```
   getRepositories: -> _.compact([@repo])
   getRepo: ->
     Grim.deprecate("Use ::getRepositories instead")
     @repo
 
+  # Public: Returns a {Promise} that resolves with either:
+  # * {Repository} if a repository can be created for the given directory
+  # * `null` if no repository can be created for the given directory.
+  #
+  # * `directory` {Directory} for which to get a {Repository}.
+  repositoryForDirectory: (directory) ->
+    path = directory.getRealPathSync()
+    promise = @directoryToRepositoryMap.get(path)
+    unless promise
+      promises = @repositoryProviders.map(
+          (provider) -> provider.repositoryForDirectory directory)
+      promise = Promise.all(promises).then((repositories) =>
```

Comment:
I actually have to make heavy use of the Preview package to figure out what is legit and what's not.

It seems like I can drop the parens but have to keep `(repositories)` on the same line as `then`.

I suppose we could log it? A provider should really be defensive and return `null`. If we don't add any handling, unhandled Promises do appear in the console, which is awesome, btw.